packages/utils/src/time.ts

```diff
@@ -1,98 +1,154 @@
-  // is not as widely supported. Namely, performance.timeOrigin is undefined in Safari as of writing.
-  if (performance.timeOrigin === undefined) {
-    // As of writing, performance.timing is not available in Web Workers in mainstream browsers, so it is not always a
-    // valid fallback. In the absence of a initial time provided by the browser, fallback to INITIAL_TIME.
-    // @ts-ignore ignored because timeOrigin is a readonly property but we want to override
-    // eslint-disable-next-line deprecation/deprecation
-    performance.timeOrigin = (performance.timing && performance.timing.navigationStart) || INITIAL_TIME;
-  }
+if (platformPerformance !== undefined) {
+  timestampSource = {
+    nowSeconds: () => (platformPerformance.timeOrigin + platformPerformance.now()) / 1000,
+  };
+}
 
-  return performance;
-})();
+/**
+ * Returns a timestamp in seconds since the UNIX epoch using the Date API.
+ */
+export const dateTimestampInSeconds = dateTimestampSource.nowSeconds.bind(dateTimestampSource);
 
 /**
- * Returns a timestamp in seconds with milliseconds precision since the UNIX epoch calculated with the monotonic clock.
+ * Returns a timestamp in seconds since the UNIX epoch using either the Performance or Date APIs, depending on the
+ * availability of the Performance API.
+ *
+ * See `usingPerformanceAPI` to test whether the Performance API is used.
+ *
+ * BUG: Note that because of how browsers implement the Performance API, the clock might stop when the computer is
+ * asleep. This creates a skew between `dateTimestampInSeconds` and `timestampInSeconds`. The
+ * skew can grow to arbitrary amounts like days, weeks or months.
+ * See https://github.com/getsentry/sentry-javascript/issues/2590.
  */
-export function timestampWithMs(): number {
-  return (crossPlatformPerformance.timeOrigin + crossPlatformPerformance.now()) / 1000;
-}
+export const timestampInSeconds = timestampSource.nowSeconds.bind(timestampSource);
+
+// Re-exported with an old name for backwards-compatibility.
+export const timestampWithMs = timestampInSeconds;
 
 /**
- * Determines if running in react native
+ * A boolean that is true when timestampInSeconds uses the Performance API to produce monotonic timestamps.
  */
-function isReactNative(): boolean {
-  return getGlobalObject<Window>().navigator?.product === 'ReactNative';
-}
+export const usingPerformanceAPI = platformPerformance !== undefined;
 
 /**
- * Performance wrapper for react native as performance.now() has been found to start off with an unusual offset.
+ * The number of milliseconds since the UNIX epoch. This value is only usable in a browser, and only when the
+ * performance API is available.
  */
-function getReactNativePerformanceWrapper(): CrossPlatformPerformance {
-  // Performance only available >= RN 0.63
+export const browserPerformanceTimeOrigin = ((): number | undefined => {
```

Comment:
I'm not sure how I feel about this tbh. Imports should imo be always deterministic, and here they are not.
WDY about writing `getBrowserTimeOrigin` instead and caching the initiall result instead? This way we'll never end up with code:

```js
import { something } from 'utils';

if (something) {
  // wait, why it can be undefined? I imported it.
}
```

On the other hand, we'll end up with the code like:

```js
import { getSometing } from 'utils';

const something = getSometing();
if (something) {
  // wait, why it can be undefined? I imported it.
}
```

I'm not against or pro any of these options, so you can decide.

Reply:
> Imports should imo be always deterministic

I agree.

Here, this value is only `undefined` if `performance` itself is `undefined`.
This is essentially replacing our polyfill (setting a `performance.timeOrigin` when it otherwise would not exist) with an explicit value we can use only within the SDK, without affecting other JS code.

The alternative was to repeat the fallback to `performance.timing.navigationStart` everywhere the `timeOrigin` is needed, which ends up being more than a one-liner  as `performance.timing` itself might be undefined (and we not using `?`... is that still the case?).

If you have yet another idea how to deal with this, I'm happy to update.